Update to trigger ID handling and behaviour for low or no hit events

brucehoward-physics opened the pull request:
An updated version for folks to start looking at. I tested that it runs the Python portion and gives sensible answers, however I still need to test the root to root conversion.

**Big big notes:**

* This updates to favour 5 over 6 where both are present, to give int32 max if no trigger is present (following a comment in my thread with Sindhu and co. noting that  1 << 31 - 1 is for things that have no external trigger (or I should say, in our case where our returned trigger array comes up with all 0 entries, which I think is what that corresponds to), and to otherwise return the first entry of the trigger array.
* The implementation here is going to be different than SPINE's but meant to provide the same end result, so I think we can just check with them or ask Noë or someone else to check the trigger matching to ensure that our final trigger product matches the one they have e.g.
* I think this method as implemented may not be quite what is wanted for FSD, based on Matt's discussion in the group Slack chat I was in. However, since there are no beam triggers where we have to ensure analyzers get the right POT, maybe this can be seen as less concerning. But worth thinking about.

-

* This also updates the behaviour for events with basically no hits: my understanding from the few quick Slack chats is that SPINE happily returns empty events. I think this is probably what you want anyway for the case of triggered events, where there not being info is actually meaningful in itself. So, this PR drastically changes behaviour with empty events. Will be good to have folks thoughts on this.
* I had noticed an event complaining about one of the time variables when not skipping "bad events" so for now just hard code the time for them to be -5. Feel free to suggest other behaviours.

-

* It's interesting that we're explicitly setting run and subrun IDs to 0 -- for data at least, can we not get something like a run and/or subrun number from the data file? 

-

* For MC right now: I have it just return trigger type = 0. Maybe someone knows if that field is actually meant to exist in 2x2 sim however?

-

* I have NOT yet checked the behaviour of some of the MC truth related bits on empty/nearly empty events in sim, e.g. I also have not checked how such empty events are handled in LArRecoND in either data or sim. **This should be part of testing. I can let people know what I see when I run a test.**

jback08 reviewed (commented):
Thanks for updating the code to process the multiple trigger IDs. However, there are problems running this script on MinRun6.3 MC files which need to be addressed.

Empty events will not fill any hierarchy ntuple output, since there will be no reconstructed PFOs to loop over; the vectors storing all of the ntuple variables will be empty.

I agree it would be good to add the run info, but I don't know where that information is stored in the H5 files; the runIDs are not listed in the [2x2 data definitions](https://github.com/DUNE/2x2_sim/wiki/File-data-definitions)

jback08 commented on ndlarflow/h5_to_root_ndlarflow.py:
```diff
@@ -86,44 +86,72 @@ def main(argv=None):
+                else:
+                    triggerIDs[i] = triggerIDsAll[i][0]
+        else:
+            triggerIDs = np.zeros( eventsToRun, dtype='int32' )
```
Does `charge/ext_trigs` also exist for MC? For CAFs, I think -1 is used to set the "All" trigger flag if nothing is available; see the comment in ND_CAFMaker's [IRecoBranchFiller](https://github.com/DUNE/ND_CAFMaker/blob/main/src/reco/IRecoBranchFiller.h#L66)

brucehoward-physics replied:
I'll check one of the 2x2 sim files you pointed me to on Slack. I'd be fine setting it to -1 otherwise.

brucehoward-physics replied:
Noting here for the record that in a 2x2 sim file pointed to me by John and in one of our single neutrino spills for NDLAr, this branch does exist and sees to be always `1`

jback08 replied:
For MC, do we want to change `triggerIDs = np.zeros( eventsToRun, dtype='int32' )` to use `np.ones` instead?

brucehoward-physics replied:
yeah we can either change that or just let it grab the 1s from the hdf5 and get rid of that block entirely?

jback08 replied:
OK, please modify this to use the hdf5 info for the MC as well, which I gather simply means removing the if statement.

brucehoward-physics commented on ndlarflow/h5_to_root_ndlarflow.py:
```diff
@@ -86,44 +86,72 @@ def main(argv=None):
             if ievt%10==0:
                 print('Currently on',ievt,'of',eventsToRun)
             event = events[ievt]
             event_calib_prompt_hits=flow_out["charge/events/","charge/calib_"+promptKey+"_hits", events["id"][ievt]]
```
Does `"charge/calib_final_hits"` exist in the file?

jback08 replied:
No, I thought the example file had both prompt & final hits, but it doesn't have final hits (the `calib_final_hits` array doesn't exist), so this explains what happened here.

jback08 commented on ndlarflow/h5_to_root_ndlarflow.py:
```diff
@@ -86,44 +86,72 @@ def main(argv=None):
-                print('This event has < 2 hit IDs, skipping')
-                continue
+                print('This event has < 2 hit IDs, setting as bad event. Trigger type (',triggerIDs[ievt],')')
+                badEvt=True
```
OK, so we could leave in empty events in the H5-to-ROOT file, but they won't produce any Pandora reco output. If we do have empty events, are all of the `other_dict` dictionary variables defined and set appropriately?

brucehoward-physics replied:
This needs to be checked and understood. I'm also not sure if in some cases the timing info might actually be sensible, like maybe the t_start is sensible for ext triggers that are empty and just is not sensible for non-triggered events. Other folks more involved to date with 2x2 data would be better suited to know this than me, but it's something I can imagine.

brucehoward-physics commented:
Time seems to not complain if we check for presence of a trigger (rather than checking `badEvt`). I propose we should check with 2x2 data folks to make sure this is expected, but it makes sense.

I also added some protections to the MC info in the `badEvt` case.

Still if we want to do this, we should probably verify if any changes at root to root conversion or `PandoraInterface` could be needed.

jback08 commented:
For the bad events, we get a warning in the ROOT conversion step:

```
root [1] rootToRootConversion(true, "MiniRunF6.3_1E19_FHC.flow.0000001.FLOW.hdf5_hits_uproot.root", "MiniRunF6.3_1E19_FHC.flow.0000001.FLOW.hdf5_hits.root")
Loaded in tree with 359 entries.
0:54:32764:-1837180944
0:0:65:1
WARNING!!! For R0, S0, E115 -- the summed number of matches expected (1) does **NOT** match the size of the packetFrac vector (0). This is a sign of trouble.
0:0:139:1
WARNING!!! For R0, S0, E156 -- the summed number of matches expected (1) does **NOT** match the size of the packetFrac vector (0). This is a sign of trouble.
0:0:183:0
DONE!
```

and in `PandoraInterface` we need to skip the events in `ProcessSPEvents()` when `nSP < 2`, otherwise it aborts when trying to create their CaloHits; the events have one ntuple entry due to the trigger/event time, but there are no hits available. I could add this extra check to #31.

brucehoward-physics commented:
* I think the warnings related to the matches count not being expected in bad events was due to not writing a matches = [0] array (as a bug). Added this in for the bad events cases.

* Moved the trigger retrieval such that it is handled the same between data and MC since MC seems to have set this value

* Added a small fix to rootToRoot conversion, where the first time it printed run/subrun/event was before it had grabbed this info from the input tree.

MariaBrigida commented:
I am happy with these changes - thanks Bruce!

jback08 commented:
The script & ROOT conversion works OK for example 2x2 MC & data files, with no warnings nor errors. So this is ready to be merged.
